Source/JavaScriptCore/wasm/WasmSectionParser.cpp

```diff
@@ -748,8 +753,26 @@ auto SectionParser::parseInitExpr(uint8_t& opcode, uint64_t& bitsOrImportNumber,
     }
 
     uint8_t endOpcode;
-    WASM_PARSER_FAIL_IF(!parseUInt8(endOpcode), "can't get init_expr's end opcode");
-    WASM_PARSER_FAIL_IF(endOpcode != OpType::End, "init_expr should end with end, ended with ", endOpcode);
+    // Don't consume the opcode byte unless it's an End so that the extended
+    // parsing mode below can consume it if needed.
+    WASM_PARSER_FAIL_IF(offset() >= length(), "can't get init_expr's end opcode");
+    endOpcode = source()[offset()];
+
+    if (endOpcode == OpType::End) {
+        m_offset++;
+        isExtendedConstantExpression = false;
+        return { };
+    }
+    WASM_PARSER_FAIL_IF(!Options::useWebAssemblyExtendedConstantExpressions(), "init_expr should end with end, ended with ", endOpcode);
+
+    // If an End doesn't appear, we have to assume it's an extended constant expression
+    // and use the full Wasm expression parser to validate.
+    size_t initExprOffset;
+    WASM_FAIL_IF_HELPER_FAILS(parseExtendedConstExpr(source() + initialOffset, length() - initialOffset, initialOffset + m_offsetInSource, initExprOffset, m_info, expectedType));
+    m_offset += (initExprOffset - (m_offset - initialOffset));
+    WASM_PARSER_FAIL_IF(!m_info->constantExpressions.tryConstructAndAppend(source() + initialOffset, initExprOffset), "could not allocate memory for init expr");
```

Comment:
The section parser communicates the constant expressions that need to be evaluated to the module record by storing them in the module info here.

In theory, we could just supply a offset/size pair into the original module bytes buffer instead of copying the bytes. However, it's not really easy to access that buffer from the module record code because neither `WebAssemblyModuleRecord` nor `WasmModule` have a reference to it.